src/accumulatormap.cpp

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2017 The PIVX developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "accumulatormap.h"
+#include "accumulators.h"
+#include "main.h"
+#include "txdb.h"
+#include "libzerocoin/Denominations.h"
+
+using namespace libzerocoin;
+using namespace std;
+
+AccumulatorMap::AccumulatorMap()
+{
+    //construct accumulators for all denominations
+    for (auto& denom : zerocoinDenomList) {
+        unique_ptr<Accumulator> uptr(new Accumulator(Params().Zerocoin_Params(), denom));
+        mapAccumulators.insert(make_pair(denom, std::move(uptr)));
+    }
+}
+
+void AccumulatorMap::Reset()
+{
+    mapAccumulators.clear();
+    for (auto& denom : zerocoinDenomList) {
+        unique_ptr<Accumulator> uptr(new Accumulator(Params().Zerocoin_Params(), denom));
+        mapAccumulators.insert(make_pair(denom, std::move(uptr)));
+    }
+}
+
+bool AccumulatorMap::Load(uint256 nCheckpoint)
+{
+    for (auto& denom : zerocoinDenomList) {
+        uint32_t nChecksum = ParseChecksum(nCheckpoint, denom);
+
+        CBigNum bnValue;
+        if (!zerocoinDB->ReadAccumulatorValue(nChecksum, bnValue)) {
+            LogPrintf("%s : cannot find checksum %d", __func__, nChecksum);
+            return false;
+        }
+
+        mapAccumulators.at(denom)->setValue(bnValue);
+    }
+    return true;
+}
+
+bool AccumulatorMap::Accumulate(PublicCoin pubCoin, bool fSkipValidation)
+{
+    CoinDenomination denom = pubCoin.getDenomination();
+    if (denom == CoinDenomination::ZQ_ERROR)
+        return false;
+
+    if (fSkipValidation)
+        mapAccumulators.at(denom)->increment(pubCoin.getValue());
+    else
+        mapAccumulators.at(denom)->accumulate(pubCoin);
+    return true;
+}
+
+CBigNum AccumulatorMap::GetValue(CoinDenomination denom)
+{
+    if (denom == CoinDenomination::ZQ_ERROR)
+        return CBigNum(0);
+    return mapAccumulators.at(denom)->getValue();
+}
+
+uint256 AccumulatorMap::GetCheckpoint()
+{
+    uint256 nCheckpoint;
+    for (auto& denom : zerocoinDenomList) {
```

Comment:
I think we should check that the # of denoms < 8 since 8*32 is 256 and the nCheckpoint would overflow 256 bits if more than 8

Reply:
You mean check that `zerocoinDenomlist` is 8? If so, I would think that is unnecessary since it is a const, but I suppose it could gaurd against the situation where we unwittingly change that definition without realizing the impact.
```
const std::vector<CoinDenomination> zerocoinDenomList = {ZQ_ONE, ZQ_FIVE, ZQ_TEN, ZQ_FIFTY, ZQ_ONE_HUNDRED, ZQ_FIVE_HUNDRED, ZQ_ONE_THOUSAND, ZQ_FIVE_THOUSAND};
```

Reply:
Yes, I meant have an assert somewhere because I think if we add another denom, we may forget about the limitation and have issues